lib/connections/kcp_dial.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package connections
+
+import (
+	"crypto/tls"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/yamux"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/xtaci/kcp-go"
+)
+
+const kcpPriority = 50
+
+func init() {
+	factory := &kcpDialerFactory{}
+	for _, scheme := range []string{"kcp", "kcp4", "kcp6"} {
+		dialers[scheme] = factory
+	}
+}
+
+type kcpDialer struct {
+	cfg    *config.Wrapper
+	tlsCfg *tls.Config
+}
+
+func (d *kcpDialer) Dial(id protocol.DeviceID, uri *url.URL) (internalConn, error) {
+	uri = fixupPort(uri, 22020)
+
+	// Try to dial via an existing listening connection
+	// giving better changes punching through NAT.
+	f := getDialingFilter()
+	var conn *kcp.UDPSession
+	var err error
+	if f != nil {
+		conn, err = kcp.NewConn(uri.Host, nil, 0, 0, f.NewConn(20, &kcpConversationFilter{}))
+		// We are piggy backing on a listener connection, no need for keepalives.
+		// Futhermore, keepalives just send garbage, which will flip our filter.
```

Comment:
I'll change this to be more understandable, but what I meant was that it will cause us to receive packets that we cannot attribute to any filter we have set up which will cause them be logged/dropped.